File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/UI/ProfileViewAdapter.swift
import PresentationCommon
import PresentationNavigation
import SwiftUI

public struct ProfileViewAdapter: View {
    @StateObject private var headerViewModel: ProfileHeaderViewModel
    @StateObject private var detailsViewModel: ProfileDetailsViewModel
    @StateObject private var woofViewModel: ProfileWoofViewModel
    @StateObject private var messageViewModel: ProfileMessageViewModel
    @StateObject private var moderationViewModel: ProfileModerationViewModel
    @Environment(\.navigator) private var navigator

    private let page: ProfilePageUIModel

    public init(page: ProfilePageUIModel) {
        self.page = page
        _headerViewModel = .init(arg1: page.domain)
        _detailsViewModel = .init(arg1: page.domain)
        _woofViewModel = .init(arg1: page.domain)
        _messageViewModel = .init(arg1: page.domain)
        _moderationViewModel = .init(arg1: page.domain)
    }

    public var body: some View {
        ProfileViewScreen(
            page: page,
            headerState: headerViewModel.state,
            detailsState: detailsViewModel.state,
            woofState: woofViewModel.state,
            messageState: messageViewModel.state,
            moderationState: moderationViewModel.state,
            onBackTap: navigator.back,
            onFavoriteTap: headerViewModel.onFavoriteTap,
            onOverflowExpandedChange: headerViewModel.onOverflowExpandedChange,
            onOverflowItemTap: onOverflowItemTap,
            onWoofTap: woofViewModel.onWoofTap,
            onWoofToastDismiss: woofViewModel.onToastDismiss,
            onMessageTextChange: messageViewModel.onTextChange,
            onMessageSendTap: messageViewModel.onSendTap,
            onMessageToastDismiss: messageViewModel.onToastDismiss,
            onModerationDialogConfirm: moderationViewModel.onDialogConfirm,
            onModerationDialogDismiss: moderationViewModel.onDialogDismiss,
            onModerationToastDismiss: moderationViewModel.onToastDismiss
        )
        .errorAdapter(
            sources: [
                ErrorSource(headerViewModel),
                ErrorSource(detailsViewModel),
                ErrorSource(woofViewModel),
                ErrorSource(messageViewModel),
                ErrorSource(moderationViewModel),
            ],
            errorMapper: ProfileErrorToToastMapper(name: page.name)
        )
        .onAppear {
            headerViewModel.onViewAppear()
            detailsViewModel.onViewAppear()
            woofViewModel.onViewAppear()
            messageViewModel.onViewAppear()
            moderationViewModel.onViewAppear()
        }
    }

    private func onOverflowItemTap(_ item: ProfileOverflowMenuItemUIModel) {
        headerViewModel.onOverflowExpandedChange(false)
        switch item {
        case .report: moderationViewModel.onReportTap()
        case .block: moderationViewModel.onBlockTap()
        }
    }
}

File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/ui/ProfileViewAdapter.kt
package com.perrystreet.woof.presentation.profile.ui

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import com.perrystreet.woof.presentation.common.error.ErrorAdapter
import com.perrystreet.woof.presentation.navigation.LocalNavigator
import com.perrystreet.woof.presentation.profile.ui.extensions.ProfileErrorToToastMapper
import com.perrystreet.woof.presentation.profile.uimodel.ProfileOverflowMenuItemUIModel
import com.perrystreet.woof.presentation.profile.uimodel.ProfilePageUIModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfileDetailsViewModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfileHeaderViewModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfileMessageViewModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfileModerationViewModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfileWoofViewModel
import org.koin.androidx.compose.koinViewModel
import org.koin.core.parameter.parametersOf

@Composable
fun ProfileViewAdapter(
    page: ProfilePageUIModel,
    headerViewModel: ProfileHeaderViewModel = koinViewModel(key = "profile.header:${page.id}") { parametersOf(page.domain) },
    detailsViewModel: ProfileDetailsViewModel = koinViewModel(key = "profile.details:${page.id}") { parametersOf(page.domain) },
    woofViewModel: ProfileWoofViewModel = koinViewModel(key = "profile.woof:${page.id}") { parametersOf(page.domain) },
    messageViewModel: ProfileMessageViewModel = koinViewModel(key = "profile.message:${page.id}") { parametersOf(page.domain) },
    moderationViewModel: ProfileModerationViewModel = koinViewModel(key = "profile.moderation:${page.id}") { parametersOf(page.domain) },
) {
    val navigator = LocalNavigator.current
    val headerState by headerViewModel.state.subscribeAsState()
    val detailsState by detailsViewModel.state.subscribeAsState()
    val woofState by woofViewModel.state.subscribeAsState()
    val messageState by messageViewModel.state.subscribeAsState()
    val moderationState by moderationViewModel.state.subscribeAsState()

    LaunchedEffect(Unit) {
        headerViewModel.onViewAppear()
        detailsViewModel.onViewAppear()
        woofViewModel.onViewAppear()
        messageViewModel.onViewAppear()
        moderationViewModel.onViewAppear()
    }

    ProfileViewScreen(
        page = page,
        headerState = headerState,
        detailsState = detailsState,
        woofState = woofState,
        messageState = messageState,
        moderationState = moderationState,
        onBackTap = navigator::back,
        onFavoriteTap = headerViewModel::onFavoriteTap,
        onOverflowExpandedChange = headerViewModel::onOverflowExpandedChange,
        onOverflowItemTap = { item ->
            headerViewModel.onOverflowExpandedChange(false)
            when (item) {
                ProfileOverflowMenuItemUIModel.Report -> moderationViewModel.onReportTap()
                ProfileOverflowMenuItemUIModel.Block -> moderationViewModel.onBlockTap()
            }
        },
        onWoofTap = woofViewModel::onWoofTap,
        onWoofToastDismiss = woofViewModel::onToastDismiss,
        onMessageTextChange = messageViewModel::onTextChange,
        onMessageSendTap = messageViewModel::onSendTap,
        onMessageToastDismiss = messageViewModel::onToastDismiss,
        onModerationDialogConfirm = moderationViewModel::onDialogConfirm,
        onModerationDialogDismiss = moderationViewModel::onDialogDismiss,
        onModerationToastDismiss = moderationViewModel::onToastDismiss,
    )

    ErrorAdapter(
        viewModels = listOf(headerViewModel, detailsViewModel, woofViewModel, messageViewModel, moderationViewModel),
        errorMapper = ProfileErrorToToastMapper(name = page.name),
    )
}
